src/Nethermind/Nethermind.Baseline/Tree/BaselineTreeTracker.cs

```diff
@@ -13,53 +13,137 @@
+        public BaselineTreeTracker(
+            Address address,
+            BaselineTree baselineTree, 
+            IBlockProcessor blockProcessor,
+            IBaselineTreeHelper baselineTreeHelper,
+            IBlockFinder blockFinder)
         {
             _address = address ?? throw new ArgumentNullException(nameof(address));
             _baselineTree = baselineTree ?? throw new ArgumentNullException(nameof(baselineTree));
-            _logFinder = logFinder ?? throw new ArgumentNullException(nameof(logFinder));
+            _blockProcessor = blockProcessor ?? throw new ArgumentNullException(nameof(blockProcessor));
+            _baselineTreeHelper = baselineTreeHelper ?? throw new ArgumentNullException(nameof(baselineTreeHelper));
             _blockFinder = blockFinder ?? throw new ArgumentNullException(nameof(blockFinder));
 
-            _timer = InitTimer();
+            StartTracking();
         }
-        
-        private Timer InitTimer()
+
+        private void StartTracking()
         {
-            Timer timer = new Timer();
-            timer.Interval = 1000;
-            timer.Elapsed += TimerOnElapsed;
-            timer.AutoReset = false;
-            return timer;
+            BlockParameter fromBlockParameter = new BlockParameter(0L);
+            if (_baselineTree.LastBlockDbHash != Keccak.Zero)
+            {
+                var lastBaselineTreeParameter = new BlockParameter(_baselineTree.LastBlockDbHash);
+                var searchResult = _blockFinder.SearchForHeader(lastBaselineTreeParameter);
+                if (!searchResult.IsError)
+                {
+                    fromBlockParameter = lastBaselineTreeParameter;
+                }
+            }
+
+            var toBlockParameter = BlockParameter.Latest;
+            _currentBlockHeader = _blockFinder.SearchForHeader(null).Object;
+            _baselineTreeHelper.BuildTree(_baselineTree, _address, fromBlockParameter, toBlockParameter);
+            _blockProcessor.BlockProcessed += OnBlockProcessed;
         }
-        
-        private void TimerOnElapsed(object sender, ElapsedEventArgs e)
+
+        private void OnBlockProcessed(object? sender, BlockProcessedEventArgs e)
+        {
+            if (_currentBlockHeader != null && _currentBlockHeader.Hash != e.Block.ParentHash)
+            {
+                if (_currentBlockHeader.Number < e.Block.Number)
+                {
+                    _baselineTreeHelper.BuildTree(_baselineTree, _address, new BlockParameter(_currentBlockHeader.Hash), new BlockParameter(e.Block.Hash));
+                }
+                else
+                {
+                    Reorganize(e.TxReceipts, e.Block.Number);
+                }
+
+                _currentBlockHeader = e.Block.Header;
+                return;
+            }
+
+            _currentBlockHeader = e.Block.Header;
+            AddFromCurrentBlock(e.TxReceipts, e.Block.Number);
```

Comment:
when you reorganize, you can still AddFromCurrentBlock here, will make code easier, unless its hashing optimazation for now.